### PR TITLE
TFAC-68 Allow Users to Delete Account

### PR DESCRIPTION
**Pull request by jedtan**

### CODE REVIEW

File: web/src/js/authentication/__tests__/user.test.js [hunk 1]
```diff
@@ -580,3 +582,27 @@ describe('updateUserEmail tests', () => {
...
+      isAnonymous: false,
+      emailVerified: true,
+      getIdToken: jest.fn(() => 'fake-token-123'),
+      delete: mockDeleteUser,
+    })
+
+    const deleteUser = require('js/authentication/user').deleteUser
+    const response = await deleteUser()
+
+    // expect(DeleteUserMutation).toHaveBeenCalled()
```
Comment (jedtan) [inline web/src/js/authentication/__tests__/user.test.js:604]: Why is this not working?

Comment (kmjennison) [reply web/src/js/authentication/__tests__/user.test.js:604]: When calling `jest.resetModules()` between tests, as we do here in the `afterEach` test teardowns, ES6 imports won't work. Instead, you have to require modules within tests. Add this:
`const DeleteUserMutation = require('js/mutations/DeleteUserMutation').default`

File: web/src/js/authentication/__tests__/user.test.js [hunk 2]
```diff
@@ -7,11 +7,13 @@ import {
   constructUrl,
   enterUsernameURL,
 } from 'js/navigation/navigation'
+import DeleteUserMutation from 'js/mutations/DeleteUserMutation'
```
Comment (kmjennison) [inline web/src/js/authentication/__tests__/user.test.js:10]: Remove (see comment below).

File: web/src/js/components/Settings/Account/__tests__/AccountComponent.test.js [hunk 3]
```diff
@@ -349,10 +350,97 @@ describe('Account component', () => {
...
+
+    deleteUserDialog
+      .find(Button)
+      .last()
+      .simulate('click')
+
+    await flushAllPromises()
+
+    expect(deleteUser).toHaveBeenCalled()
+    expect(replaceUrl).toHaveBeenCalledWith(loginURL)
```
Comment (kmjennison) [inline web/src/js/components/Settings/Account/__tests__/AccountComponent.test.js:444]: Test broken: add second arg.

File: web/src/js/authentication/__tests__/user.test.js [hunk 4]
```diff
@@ -580,3 +582,27 @@ describe('updateUserEmail tests', () => {
     })
   })
 })
+
+describe('deleteUser tests', () => {
+  it('works as expected', async () => {
```
Comment (kmjennison) [inline web/src/js/authentication/__tests__/user.test.js:587]: Add `expect.assertions(3)` at the top. For async tests, this ensures it actually gets through every assertion you'd expect it to.

File: web/src/js/components/Authentication/Authentication.js [hunk 5]
```diff
@@ -394,6 +395,16 @@ class Authentication extends React.Component {
                   <span>First, please login again.</span>
                 </Typography>
               ) : null}
+              {urlParams.accountDeleted ? (
```
Comment (kmjennison) [inline web/src/js/components/Authentication/Authentication.js:398]: Add test

**Review by kmjennison — CHANGES_REQUESTED** [review]
Nice work! After these changes, please deploy to dev for manual testing.

File: web/src/js/authentication/__tests__/user.test.js (hunk 1 above)
Comment (jedtan) [reply web/src/js/authentication/__tests__/user.test.js:604]: Ah! Thanks

File: web/src/js/authentication/user.js [hunk 6]
```diff
@@ -345,3 +347,34 @@ export const updateUserEmail = async newEmail => {
...
+        DeleteUserMutation(
+          environment,
+          user.id,
+          () => {},
+          response => {
+            console.error('Error deleting user:')
+            logger.error(response)
+          }
+        )
+        firebase
```
[the remark this answers is not in the document]
Comment (jedtan) [reply web/src/js/authentication/user.js:364]: Good catch.

**Review by kmjennison — APPROVED** [review]
LGTM! 🦊

**Pull request merged**